### DISCUSSION TITLE
How to resovle that some properties and events removed from BootStrapperApplication when upgrading from WiX3 to Wix 4

### DISCUSSION THREAD

**Discussion by Federau** [wixtoolset/issues#7718] [Questions] [answered]
Hi guys:
     I am doing a migration for my bundle project from Wix3 to Wix4. I have a custom BootstrapperApplication.  currently I encounter the following issues:
1.Command Property was removed 
```
		bool CheckIfGUIMode()
		{
			return Command.Display == Display.Passive || Command.Display == Display.Full;
		}
```
2. ResolveSource event was removed 
```
		{
			ResolveSource += OnResolveSource;
			RunQuietMode();
			Engine.CloseSplashScreen();
		}
```

3.Installed property of  DetectBeginEventArgs was removed
```
		void OnDetectBegin(object sender, DetectBeginEventArgs e)
		{
			if (e.Installed)
			{
				ProductInstallationState = ProductInstallationState.SameVersionInstalled;
			}
			else
			{
				ProductInstallationState = ProductInstallationState.NotInstalled;
			}
		}
```
4.Operation propert in DetectRelatedBundleEventArgs was removed but found in DetectRelatedMsiPacakageEventArgs
5 Result property in PlanRelatedBundleEventArgs was remvoed.
```
#pragma warning disable CA2109 // Review visible event handlers
		public void OnPlanRelatedBundle(object sender, PlanRelatedBundleEventArgs e)
#pragma warning restore CA2109 // Review visible event handlers
		{
			if (e != null)
			{
				if (PlannedAction != LaunchAction.Install)
				{
					return;
				}

				if (ProductInstallationState == ProductInstallationState.NewerVersionInstalled)
				{
					e.Result = Result.Cancel;
					e.State = RequestState.Present;
					LogError($"Newer version of this bundle is already installed. Setup will exit now.");
				}

				if (ProductInstallationState == ProductInstallationState.SameVersionInstalled)
				{
					e.State = RequestState.Absent;
				}
			
```
6.there is a IEngine parameter in constructor, I added the parameter in ApplicationBootstrapper too,but where to inject it. should I crate a Engine for that
```
		public ApplicationBootstrapper(IEngine engine) : base(engine)
		{
			PlanRelatedBundle += OnPlanRelatedBundle;
			DetectRelatedBundle += OnDetectRelatedBundle;
			DetectBegin += OnDetectBegin;
		}
```

**Comment by Federau** [comment]
I think I got some idea here...   (https://github.com/orgs/wixtoolset/discussions/7591)

**Comment by BMurri** [comment, marked as answer]
A combination of [Converting bundles](https://wixtoolset.org/docs/fourthree/faqs/#converting-bundles) and [WixToolset.WixBA](https://github.com/wixtoolset/wix/tree/develop/src/test/burn/WixToolset.WixBA) should get you on the right track to convert your BA from v3 to v4 (at least it did for me :wink:)

**Reply by Federau** [reply]
I'll give it shot, thank you !

**Comment by Federau** [comment]
Actually I am a little confused about this wixBA project. since wix4 no longer supports net5 and pervious https://wixtoolset.org/docs/fourthree/faqs/#upgrading-v3-managed-bootstrapper-applications 
why do we need to add net472 framework target inside. Can we use net6-windows directly?

**Reply by BMurri** [reply]
You have options. You can go the "modern" route and do Net-6+ via `WixDotNetCoreBootstrapperApplicationHost` (no framework needed) or you can maintain compatibility with your custom actions and stay with Net-Framework via `WixManagedBootstrapperApplicationHost` (no framework supported). Note that WiX v4 does NOT support for non-Framework managed custom actions.

The biggest advantage of staying with the Framework is that you usually don't have to preinstall anything on almost every machine. The biggest disadvantage of going "modern" is that support for single-file BAs is broken by design by the dotnet team (although Net-8 AOT looks interesting).

**Reply by Federau** [reply]
so if I have a custom action project which depends on Wix3 ,Then should I migrate action project to Wix4 first?
